### PR TITLE
Inline an account typeahead for Complex Vouchers

### LINKED ISSUE #1289: (proposal) use typeahead for defining row account(s) in complex vouchers
I propose that we re-factor the complex vouchers module to use typeaheads to select accounts. This would be similar in funcionality to the patient invoice module where inventory items can be searched for and then 'locked in' by selecting the item or pressing enter. 

Pros: 
* The current `ui-select` that is used is too tall for our grid rows. This is a bad user experience with some of the select missing from the bottom. A typeahead avoids this by being used on a simple native input.

* The DOM for the `ui-select` is much larger and more complex. This means that if there are a large number of rows in a voucher every row will have to render this complex element. It will also have to be rendered and removed when scrolling with `ui-grid`. A typeahead would simply be an input and it would be removed once the user has 'locked in' their choice. 
* A uniform `ui-grid` entity selection (accounts, inventory items). 
* `ui-select` overrides the tab key and forces you to tab through all of the options, in this case this is not helpful as it is just a small element on a complex page.
* If there are many accounts it can be much more convenient to just type in the number vs. given the display options in the first place. 

Cons: 
* `ui-select` is used for account select throughout the application, this would be a different experience however it would be inline with how selects work in grids. 
* If you have completely forgotten an account, both the name and the account number you will have to go and look it up in the chart of accounts.

### PR DESCRIPTION
**Pull request by jniles**
This commit adds the bhAccountTypeaheadInlineComponent for use in grids,
such as the Complex Voucher grid.  It is prototyped and demonstrated in
the Complex Vouchers module.

Closes #1289.

### CODE REVIEW

File: client/src/js/components/bhAccountTypeaheadInline.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+angular.module('bhima.components')
+  .component('bhAccountTypeaheadInline', {
```
Comment (mbayopanda) [inline client/src/js/components/bhAccountTypeaheadInline.js:2]: 👍

File: client/src/modules/vouchers/complex-voucher.ctrl.js [hunk 2]
```diff
@@ -174,9 +174,9 @@ function ComplexJournalVoucherController(
   }
 
   /** Get the selected currency symbol */
-  function currencySymbol(currency_id) {
-    if (!currency_id) { return; }
-    return Currencies.symbol(currency_id);
+  function currencySymbol(currencyId) {
```
Comment (mbayopanda) [inline client/src/modules/vouchers/complex-voucher.ctrl.js:177]: 👍

File: client/src/modules/vouchers/templates/account.grid.tmpl.html [hunk 3]
```diff
@@ -1,20 +1,7 @@
...
-     class="ui-grid-ui-select"
-     ui-select-focus-patch
-     ui-disable-choice="account.type_id === grid.appScope.bhConstants.accounts.TITLE"
-     repeat="account.id as account in grid.appScope.Voucher.accounts | filter: { 'hrlabel' : $select.search } track by account.id">
-     <strong ng-bind-html="account.number | highlight:$select.search"></strong>
-     <span ng-bind-html="account.label | highlight:$select.search"></span>
-   </ui-select-choices>
- </ui-select>
+<div class="ui-grid-cell-contents">
+  <bh-account-typeahead-inline
```
Comment (mbayopanda) [inline client/src/modules/vouchers/templates/account.grid.tmpl.html:2]: 👍

File: test/data.sql [hunk 4]
```diff
@@ -14,7 +14,7 @@ INSERT INTO `project` VALUES (1,'Test Project A','TPA',1,1,0),(2,'Test Project B
 
 -- Accounts
 INSERT INTO `account` VALUES
-  (3626,6,1,1000,'Test Capital Account',0,0,NULL,NULL,'2015-11-04 13:25:12',1,NULL,NULL,NULL,1,1),
+  (3626,3,1,1000,'Test Capital Account',0,0,NULL,NULL,'2015-11-04 13:25:12',1,NULL,NULL,NULL,1,1),
```
Comment (mbayopanda) [inline test/data.sql:17]: 💃

File: test/end-to-end/vouchers/complex.spec.js [hunk 5]
```diff
@@ -6,30 +6,42 @@ const ComplexVoucherPage = require('./complex.page');
 const FU = require('../shared/FormUtils');
 const GU = require('../shared/GridUtils');
 
-describe('Complex Vouchers', function () {
+describe('Complex Vouchers', () => {
   before(() => helpers.navigate('vouchers/complex'));
 
-  it('creates a complex voucher', () => {
+  it('creates a complex voucher', function () {
```
Comment (mbayopanda) [inline test/end-to-end/vouchers/complex.spec.js:12]: I'm just curious to know why you replaced the `() => ` by a `function ()`, is there any reason ?

**Review by mbayopanda — APPROVED** [review]
LGTM

File: test/end-to-end/vouchers/complex.spec.js (hunk 5 above)
Comment (jniles) [reply test/end-to-end/vouchers/complex.spec.js:12]: Yeah - arrow functions don't have a `this` object inside them, and on line 16 I use `this.timeout()`.

[According to stack overflow](https://stackoverflow.com/a/15982893), you are supposed to be able to use `it('something', () => { /* ... */ }).timout(45000);`.   That version didn't work for me though... I think our mocha version might not be up to date.

For more information on setting the timeout inside a test, [you can check out Mocha's documentation on it](http://mochajs.org/#test-level).

**Pull request merged**